presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFunction.java

```diff
@@ -0,0 +1,236 @@
+                }
+            }
+        }
+        else if (javaType == Slice.class) {
+            for (Object e : elements) {
+                if (e == null) {
+                    elementsBlockBuilder.appendNull();
+                } else {
+                    type.writeSlice(elementsBlockBuilder, utf8Slice((String) e));
+                }
+            }
+        }
+        else {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Unexpected type %s", javaType.getName()));
+        }
+
+        return elementsBlockBuilder.build();
+    }
+
+    private static List<Object> blockToList(final Type type, Block block)
+    {
+        int positionCount = block.getPositionCount();
+        ArrayList<Object> result = new ArrayList<>(positionCount);
+
+        Class<?> javaType = type.getJavaType();
+        if (javaType == boolean.class) {
+            for (int i = 0; i < block.getPositionCount(); i++) {
+                if (block.isNull(i)) {
+                    result.add(null);
+                } else {
+                    result.add(type.getBoolean(block, i));
+                }
+            }
+        }
+        else if (javaType == long.class) {
+            for (int i = 0; i < block.getPositionCount(); i++) {
+                if (block.isNull(i)) {
+                    result.add(null);
+                } else {
+                    result.add(type.getLong(block, i));
+                }
+            }
+        }
+        else if (javaType == double.class) {
+            for (int i = 0; i < block.getPositionCount(); i++) {
+                if (block.isNull(i)) {
+                    result.add(null);
+                } else {
+                    result.add(type.getDouble(block, i));
+                }
+            }
+        }
+        else if (javaType == Slice.class) {
+            for (int i = 0; i < block.getPositionCount(); i++) {
+                if (block.isNull(i)) {
+                    result.add(null);
+                } else {
+                    result.add(type.getSlice(block, i));
+                }
+            }
+        }
+        else {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, format("Unexpected type %s", javaType.getName()));
+        }
+
+        return result;
+    }
+
+    public static Slice distinct(final Type type, Slice jsonArray)
+    {
+        List<Object> elements;
+        try {
+            elements = OBJECT_MAPPER.readValue(jsonArray.getInput(), COLLECTION_TYPE);
+        }
+        catch (IOException e) {
+            throw new PrestoException(INVALID_FUNCTION_ARGUMENT, e);
+        }
+
+        if (elements.isEmpty()) {
+            return toStackRepresentation(ImmutableList.of());
+        }
+
+        // This line of code can be removed once internal array representation becomes block
+        Block elementsBlock = listToBlock(type, elements);
+
+        Block hashBlock = TypeUtils.getHashBlock(ImmutableList.of(type), elementsBlock);
```

Comment:
I don't think we need this. GroupByHash will just compute the hashes for you, since you passed it Optional.empty()